Mathlib/LinearAlgebra/QuadraticForm/Basic.lean

```diff
@@ -500,397 +512,411 @@ end SemiringOperators
 
 @[simp]
-theorem sub_apply (Q Q' : QuadraticForm R M) (x : M) : (Q - Q') x = Q x - Q' x :=
+theorem sub_apply (Q Q' : QuadraticMap R M N) (x : M) : (Q - Q') x = Q x - Q' x :=
   rfl
-#align quadratic_form.sub_apply QuadraticForm.sub_apply
+#align quadratic_form.sub_apply QuadraticMap.sub_apply
 
-instance : AddCommGroup (QuadraticForm R M) :=
+instance : AddCommGroup (QuadraticMap R M N) :=
   DFunLike.coe_injective.addCommGroup _ coeFn_zero coeFn_add coeFn_neg coeFn_sub
     (fun _ _ => coeFn_smul _ _) fun _ _ => coeFn_smul _ _
 
 end RingOperators
 
 section Comp
 
-variable [CommSemiring R] [AddCommMonoid M] [Module R M]
-variable [AddCommMonoid N] [Module R N]
+variable [CommSemiring R] [AddCommMonoid M] [Module R M] [AddCommMonoid N] [Module R N]
+variable {N' : Type v} [AddCommMonoid N'] [Module R N']
 
-/-- Compose the quadratic form with a linear function. -/
-def comp (Q : QuadraticForm R N) (f : M →ₗ[R] N) : QuadraticForm R M where
+/-- Compose the quadratic map with a linear function on the right. -/
+def comp (Q : QuadraticMap R N' N) (f : M →ₗ[R] N') : QuadraticMap R M N where
   toFun x := Q (f x)
   toFun_smul a x := by simp only [map_smul, f.map_smul]
   exists_companion' :=
     let ⟨B, h⟩ := Q.exists_companion
     ⟨B.compl₁₂ f f, fun x y => by simp_rw [f.map_add]; exact h (f x) (f y)⟩
-#align quadratic_form.comp QuadraticForm.comp
+#align quadratic_form.comp QuadraticMap.comp
 
 @[simp]
-theorem comp_apply (Q : QuadraticForm R N) (f : M →ₗ[R] N) (x : M) : (Q.comp f) x = Q (f x) :=
+theorem comp_apply (Q : QuadraticMap R N' N) (f : M →ₗ[R] N') (x : M) : (Q.comp f) x = Q (f x) :=
   rfl
-#align quadratic_form.comp_apply QuadraticForm.comp_apply
+#align quadratic_form.comp_apply QuadraticMap.comp_apply
 
-/-- Compose a quadratic form with a linear function on the left. -/
+/-- Compose a quadratic map with a linear function on the left. -/
 @[simps (config := { simpRhs := true })]
-def _root_.LinearMap.compQuadraticForm [CommSemiring S] [Algebra S R] [Module S M]
-    [IsScalarTower S R M] (f : R →ₗ[S] S) (Q : QuadraticForm R M) : QuadraticForm S M where
+def _root_.LinearMap.compQuadraticMap [CommSemiring S] [Algebra S R] [Module S N] [Module S M]
+    [IsScalarTower S R N] [IsScalarTower S R M] [Module S N']
+    (f : N →ₗ[S] N') (Q : QuadraticMap R M N) : QuadraticMap S M N' where
   toFun x := f (Q x)
   toFun_smul b x := by simp only [Q.map_smul_of_tower b x, f.map_smul, smul_eq_mul]
   exists_companion' :=
     let ⟨B, h⟩ := Q.exists_companion
     ⟨(B.restrictScalars₁₂ S S).compr₂ f, fun x y => by
       simp_rw [h, f.map_add, LinearMap.compr₂_apply, LinearMap.restrictScalars₁₂_apply_apply]⟩
-#align linear_map.comp_quadratic_form LinearMap.compQuadraticForm
+#align linear_map.comp_quadratic_form LinearMap.compQuadraticMap
 
 end Comp
 
 section CommRing
 
-variable [CommSemiring R] [AddCommMonoid M] [Module R M]
+variable [CommSemiring R] [CommSemiring N] [AddCommMonoid M] [Module R M]
```

Comment:
It's possible to have `[NonUnitalNonAssocCommSemiring N]` here - but then `linMulLinSelfPosDef` breaks.

Reply:
This is being dealt with in #12617 - for now we will stick with the `CommSemiring` case.